### PR TITLE
[2025-11-20] Pułapki adnotacji @Transactional

### PR DESCRIPTION
**Pull request by Aezy34**

### CODE REVIEW

**Comment by sirtekyt** [comment]
- [Zaplanowany na opublikowanie 10.04.2026r.]

File: _posts/pl/2022-01-25-transactions.md [hunk 1]
```diff
@@ -274,3 +274,7 @@ public class MailService {
 
 ## Podsumowanie
 Teoretyczne wprowadzenie do systemu zarządzania transakcjami w Javie już za nami! Na początku dowiedzieliśmy się czym w ogóle są transakcje, dlaczego są ważne i co je cechuje. Następnie przedstawiony został mechanizm działania transakcji w JDBC po to, by na koniec zaprezentować uproszczenie implementacji tego mechanizmu w Springu. W kolejnym artykule z tej serii dowiemy się jakie problemy napotkaliśmy wykorzystując mechanizm zarządzania transakcjami, jak udało się go zlokalizować oraz na czym polegała poprawka.
```
Comment (martawpiatek) [inline _posts/pl/2022-01-25-transactions.md]: Brakuje przecinków:
1. Na początku dowiedzieliśmy się, czym w ogóle są transakcje, dlaczego są ważne i co je cechuje. 
2. W kolejnym artykule z tej serii dowiemy się, jakie problemy napotkaliśmy, wykorzystując mechanizm zarządzania transakcjami, jak udało się go zlokalizować oraz na czym polegała poprawka.

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    private List<Car> cars = new ArrayList<>();
+
+    // konstruktory, gettery, settery
+
+}
+```
+
+## 1. @Transactional domyślnie jest realizowany przez proxy
+Domyślnie Spring dla wszystkich klas lub metod oznaczonych adnotacją `@Transactional` tworzy proxy (dynamiczne proxy JDK albo proxy CGLIB w zależności od sytuacji oraz ustawień), co umożliwia zastosowanie logiki transakcyjnej (np. rozpoczęcia i commitowania transakcji) przed oraz po wykonaniu wywoływanej metody. Oznacza to, że adnotacja ta będzie działała tylko w przypadku metod publicznych - metody o innej widoczności po prostu zignorują tę adnotację bez żadnych ostrzeżeń, ponieważ nie są one obsługiwane przez proxy. Dodatkowo przechwytywane będą tylko zewnętrzne wywołania metod, czyli takie, które przechodzą przez proxy. Wszelkie wywołania metod wewnątrz tego samego komponentu nie spowodują rozpoczęcia transakcji, nawet jeśli metoda jest oznaczona adnotacją `@Transactional`. W przypadku, gdy chcemy, aby wewnętrzna metoda działała jednak w sposób transakcyjny możemy posiłkować się wstrzyknięciem beana do samego siebie (self-injection), co pozwala nam na użycie proxy stworzonego przez Springa.
```
Comment (martawpiatek) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: Interpunkcja:

W przypadku gdy chcemy, aby wewnętrzna metoda działała jednak w sposób transakcyjny, możemy posiłkować się wstrzyknięciem beana do samego siebie (self-injection), co pozwala nam na użycie proxy stworzonego przez Springa.

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+        1  | Volkswagen Jetta | 10000
+    */
+}
+```
+W tym przykładowym serwisie walidacja opiera się na sprawdzeniu, czy nowy przebieg samochodu nie jest niższy od starego. Jeśli nowy przebieg jest niepoprawny, to metoda `carRepository.save(car)`, która wprost zapisuje tę encję do bazy, nie zostanie wykonana. Ponieważ jednak znajdujemy się w transakcji, a encja ta znajduje się w kontekście persystencji (została wcześniej wyciągnięta z bazy za pośrednictwem `carRepository.findById(carId)`), w przypadku ustawienia niepoprawnej wartości poprzez `car.setMileage()` zmiana ta zostanie odzwierciedlona na bazie danych, nawet jeśli kod naszego serwisu nie wskazuje na to wprost.
+
+Rozwiązaniem tego problemu jest pilnowanie, aby wszelkie zmiany stanu obiektu znajdującego się w kontekście persystencji odbywały się w momencie, w którym dozwolony jest jego zapis. Jeśli z jakiegoś powodu nie jest to możliwe możemy ratować się np. odłączając obiekt z kontekstu persystencji lub korzystając z obiektu pośredniczącego, który na końcu zostanie zsynchronizowany z oryginalnym obiektem. Poniżej znajduje się przykładowa implementacja `MileageUpdater` z poprawioną logiką.
```
Comment (martawpiatek) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: Interpunkcja:

Jeśli z jakiegoś powodu nie jest to możliwe, możemy ratować się, np. odłączając obiekt z kontekstu persystencji lub korzystając z obiektu pośredniczącego, który na końcu zostanie zsynchronizowany z oryginalnym obiektem.

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+        id | model            | mileage
+        ---+------------------+--------
+        1  | Volkswagen Jetta | 197000
+    */
+}
+```
+
+### Hibernate porównuje obiekty z tym co znajduje się w kontekście persystencji
+
+W ramach transakcji `dirty checking` jest podstawowym mechanizmem, dzięki któremu Hibernate wie jakie encje muszą zostać zaktualizowane, a które tego nie wymagają. Warto mieć na uwadze, że Hibernate załaduje do kontekstu persystencji wszystkie obiekty zwrócone przez zapytania bazodanowe, nawet jeśli zwrócą one rekordy, które w praktyce nie istnieją.
```
Comment (martawpiatek) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: Interpunkcja:

W ramach transakcji `dirty checking` jest podstawowym mechanizmem, dzięki któremu Hibernate wie, jakie encje muszą zostać zaktualizowane, a które tego nie wymagają.

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+    */
+    /*
+        Stan bazy po wykonaniu tryToAddNotExistingCars():
+        > SELECT * FROM car;
+        mileage | car_owner_id | id | model | vin
+        --------+--------------+----+-------+----
+    */
+}
+```
+Zapytanie `returnNotExistingCars()` zwraca samochód, który został stworzony bezpośrednio w tym zapytaniu i nie jest zapisany w bazie danych. Metoda `tryToAddNotExistingCars()` próbuje zapisać ten samochód, jednak tabela pozostanie pusta pomimo tego, że używamy `carRepository.saveAll(cars)`. Z punktu widzenia Hibernate, nieistniejący samochód istnieje w kontekście persystencji (jest zarządzany przez Hibernate) i jego stan nie został zmieniony, dlatego nie ma potrzeby wykonania operacji `INSERT` do bazy danych. W tej sytuacji aby dodać tę encję trzeba np. odłączyć ją od kontekstu za pośrednictwem `EntityManager.detach()`.
```
Comment (martawpiatek) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: Interpunkcja:

W tej sytuacji, aby dodać tę encję, trzeba np. odłączyć ją od kontekstu za pośrednictwem `EntityManager.detach()`.

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+
+## 3. @Transactional, a checked exceptions
+Jednym z kluczowych mechanizmów transakcji jest rollback, który wycofuje wszystkie zmiany w przypadku natrafienia na wyjątek. Domyślnie mechanizm ten wyzwalany jest przy wyjątkach `unchecked exceptions` dziedziczących po `RuntimeException` oraz przy wyjątkach dziedziczących po klasie `Error`, natomiast nie działa w przypadku `checked exceptions`, czyli wyjątków dziedziczących po klasie `Exception`. Aby mechanizm ten działał także w tym przypadku, należy wprost określić wyjątki, które mają uruchamiać rollback używając parametru `rollbackFor` w adnotacji `@Transactional` - przykładowo `@Transactional(rollbackFor = SomeBusinessException.class)`
```
Comment (martawpiatek) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: Interpunkcja:

Aby mechanizm ten działał także w tym przypadku, należy wprost określić wyjątki, które mają uruchamiać rollback, używając parametru `rollbackFor` w adnotacji `@Transactional` - przykładowo: `@Transactional(rollbackFor = SomeBusinessException.class)`

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+
+## 4. @Transactional w testach
+Popularną praktyką w testach integracyjnych zahaczających o bazę danych jest użycie w nich adnotacji `@Transactional` (bezpośrednio lub pośrednio np. przez adnotację `@DataJpaTest`). Z pozoru jest to wygodne rozwiązanie, które np. gwarantuje nam czyszczenie bazy po każdym teście, jednak w praktyce może prowadzić do kilku pułapek. Istnieje ryzyko, że transakcje i ich zakresy obecne w naszych testach będą różniły się od tych obecnych w kodach produkcyjnych przez co nasze testy nie będą działały zgodnie z oczekiwaniami.
```
Comment (martawpiatek) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: Interpunkcja:

1. (bezpośrednio lub pośrednio, np. przez adnotację `@DataJpaTest`). 
2. Istnieje ryzyko, że transakcje i ich zakresy obecne w naszych testach będą różniły się od tych obecnych w kodach produkcyjnych, przez co nasze testy nie będą działały zgodnie z oczekiwaniami.

**Review by martawpiatek — CHANGES_REQUESTED** [review]

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+[Car(id=1, vin=SZUKANY, model=Toyota Yaris, mileage=199989),
+    Car(id=2, vin=INNY, model=Audi A6, mileage=199989)]
+```
+Pierwszą reakcją może być myśl, że zapytanie jest niepoprawne, jednak w tym przypadku powód jest inny. Powyższy test jest opakowany w `@Transactional` - oznacza to, że cały jego przebieg (uzupełnienie bazy danych testowymi danymi w `populateDb()` oraz sam właściwy test zapytania `shouldFindOwnerWithFilteredCars()`) wykonuje się w jednej transakcji, a tym samym w jednej sesji Hibernate. Ponieważ zapis do bazy wiąże się z wrzuceniem obiektu do kontekstu persystencji, Hibernate nie uwzględnia w pełni treści zapytania i zamiast pobrać wartości z bazy, pobiera je ze swojego cache L1. Taka sytuacja nie miałaby miejsca w przypadku kodów produkcyjnych, w których zapis do bazy oraz odczyt byłby najprawdopodobniej w zupełnie odrębnych transakcjach. 
+
+Jeśli pozbędziemy się adnotacji `@Transactional`, test będzie przechodził zgodnie z oczekiwaniami. W takim przypadku jednak musimy sami zadbać o wyczyszczenie bazy po wykonanym teście - np. używając adnotacji `@AfterEach` i metody `carRepository.deleteAll()`.
+
+#### Więcej o @Transactional
```
Comment (sirtekyt) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: W sumie może źródła dorzucimy?

File: _posts/pl/2022-01-25-transactions.md [hunk 9]
```diff
@@ -274,3 +274,7 @@ public class MailService {
 
 ## Podsumowanie
 Teoretyczne wprowadzenie do systemu zarządzania transakcjami w Javie już za nami! Na początku dowiedzieliśmy się czym w ogóle są transakcje, dlaczego są ważne i co je cechuje. Następnie przedstawiony został mechanizm działania transakcji w JDBC po to, by na koniec zaprezentować uproszczenie implementacji tego mechanizmu w Springu. W kolejnym artykule z tej serii dowiemy się jakie problemy napotkaliśmy wykorzystując mechanizm zarządzania transakcjami, jak udało się go zlokalizować oraz na czym polegała poprawka. 
+
+### Więcej o @Transactional
+
+Więcej o adnotacji _@Transactional_ można przeczytać w [Pułapki adnotacji @Transactional]({% post_url pl\2026-04-10-pulapki-adnotacji-transactional %}).
```
Comment (sirtekyt) [inline _posts/pl/2022-01-25-transactions.md]: post_url pl\2026-04-10-pulapki-adnotacji-transactional dałbym po prostu linki, bo widziałem że się sypie przez to, tak samo w drugim poście

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 10]
```diff
@@ -0,0 +1,452 @@
+---
+layout: post
+title: Pułapki adnotacji @Transactional
+description: ""
+date: 2026-04-10T08:00:00+01:00
+published: false
```
Comment (sirtekyt) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: true

**Comment by sirtekyt** [comment]
dociągnij mastera

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 11]
```diff
@@ -0,0 +1,452 @@
+---
+layout: post
+title: Pułapki adnotacji @Transactional
+description: ""
```
Comment (sirtekyt) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: dałbym 1 zdanie description,

**Review by sirtekyt — CHANGES_REQUESTED** [review]

File: _posts/pl/2022-01-25-transactions.md (hunk 9 above)
Comment (Aezy34) [reply _posts/pl/2022-01-25-transactions.md]: Teraz będzie już działać, źle tego użyłem :)

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+        assertThat(owners).hasSize(1);
+        CarOwner carOwner = owners.get(0);
+        assertThat(carOwner.getCars())
+                .hasSize(1)
+                .extracting(Car::getVin)
+                .isEqualTo(List.of("SZUKANY"));
+    }
+}
+```
+Powyższy test wypełnia bazę danych w metodzie `populateDb()`, a następnie wykonuje testowane zapytanie do bazy. Następnie robi asercje, aby upewnić się, że zwrócony został właściciel z dokładnie jednym, szukanym samochodem. Uruchomienie tego testu zakończy się jednak błędem:
```
Comment (martawpiatek) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: Aby uniknąć powtórzenia:

Powyższy test wypełnia bazę danych w metodzie `populateDb()`, po czym wykonuje testowane zapytanie do bazy. Następnie ...

**Review by martawpiatek — CHANGES_REQUESTED** [review]

File: _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md [hunk 13]
```diff
@@ -0,0 +1,458 @@
...
+        ---+------------
+    */
+
+}
+```
+Przedstawiona klasa `AspectJTransactionalService` poprawnie wycofa transakcję, pomimo że metoda `saveCar()` jest prywatna oraz jest wywoływana wewnętrznie.
+
+### 2. @Transactional, Hibernate i dirty checking 
+
+#### Hibernate aktualizuje zarządzane obiekty niewprost
```
Comment (sirtekyt) [inline _posts/pl/2026-04-10-pulapki-adnotacji-transactional.md]: nie wprost

**Review by sirtekyt — APPROVED** [review]
